### PR TITLE
Add a visual style test app

### PR DESCRIPTION
**Pull request by bremac**
See `visual_style/README.md` for documentation, and `visual_style/TODO.md` for features that weren't implemented. Right now, there is only a single style test page registered under `/visual-style/` though new components can be registered easily.

### CODE REVIEW

**Comment by bremac** [comment]
Everything in one file makes it hard to add custom scripts with the component, which means that some of the bootstrap examples can't be made to work (ex. those involving tooltips.) I think I'll have to change it so that each template renders on a separate page, with a nav bar to allow you to move between pages.

File: visual_style/management/commands/update_bootstrap_examples.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    This function does not recurse into subdirectories.
+
+    """
+
+    chunks = [
+        '{% extends "visual_style/snippet_details.html" %}',
+        '{% load static %}',
+        '{% block scripts %}',
+        '{{ block.super }}',
+        '<script src="{% static "bower/bootstrap/docs/assets/js/application.js" %}"></script>',
```
Comment (pmdarrow) [inline visual_style/management/commands/update_bootstrap_examples.py]: If this is split out into a separate app, we shouldn't assume they'll be using bower.

File: visual_style/views.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from django.http import HttpResponseRedirect
+from django.template.response import TemplateResponse
+from django.views.generic.base import View, ContextMixin
+
+from .template_finder import find_templates_in_prefix
+
+
+class VisualStyleView(View, ContextMixin):
+    """
+    """
```
Comment (pmdarrow) [inline visual_style/views.py]: Did you mean to leave some comments here?

File: visual_style/views.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    # We can't use a TemplateView because the template rendered
+    #  varies based on keyword arguments captured from the URL
+
+    template_prefix = ('visual_style', 'snippets')
+
+    def get_snippets(self, active_template_filename):
+        """
+        """
```
Comment (pmdarrow) [inline visual_style/views.py:29]: And here?

**Comment by pmdarrow** [comment]
I think this should live in it's own app/repo and not in django-base. Thoughts?

File: visual_style/views.py (hunk 3 above)
Comment (bremac) [inline visual_style/views.py:29]: Uh yeah, apparently so.

File: visual_style/management/commands/update_bootstrap_examples.py (hunk 1 above)
Comment (bremac) [inline visual_style/management/commands/update_bootstrap_examples.py]: Good point. Should we copy the static bower assets to the static directory in the app as part of extracting sample components?

**Comment by bremac** [comment]
Agreed 100%. Integrating this into repos that were derived from django-base will be a huge pain otherwise.

File: visual_style/management/commands/update_bootstrap_examples.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    visual_style_path = up_directories(__file__, count=2)
+    snippet_dirname = os.path.join("templates", "visual_style", "snippets")
+    snippet_directory = os.path.join(visual_style_path, snippet_dirname)
+    return os.path.join(snippet_directory, "bootstrap.html")
+
+
+class Command(BaseCommand):
+    help = 'Updates the bootstrap examples in the visual_style app'
+
+    def handle(self, *args, **kwargs):
```
Comment (pmdarrow) [inline visual_style/management/commands/update_bootstrap_examples.py:196]: This command LGTM but how fragile is it? Do you think it'll die when we run it against Bootstrap 3?

Comment (bremac) [inline visual_style/management/commands/update_bootstrap_examples.py:196]: Maybe. No matter what we do, pulling examples and headers is going to be tied to the markup.

File: visual_style/views.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+from django.views.generic.base import View, ContextMixin
+
+from .template_finder import find_templates_in_prefix
+
+
+class VisualStyleView(View, ContextMixin):
+    """
+    """
+
+    # We can't use a TemplateView because the template rendered
```
Comment (pmdarrow) [inline visual_style/views.py]: You should be able to inherit TemplateView if you want, try the following:

``` python
from django.views.generic import TemplateView

class DynamicTemplateView(TemplateView):

    def get_template_names(self):
        return ['%s.html' % self.kwargs['template']]
```

File: visual_style/management/commands/update_bootstrap_examples.py (hunk 1 above)
Comment (pmdarrow) [inline visual_style/management/commands/update_bootstrap_examples.py]: Hmm. What about just adding a setting for the Bootstrap path that can be overridden?

File: visual_style/views.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    #  varies based on keyword arguments captured from the URL
+
+    template_prefix = ('visual_style', 'snippets')
+
+    def get_snippets(self, active_template_filename):
```
Comment (bremac) [inline visual_style/views.py:21]: I don't even know if this logic belongs in a view...

Comment (pmdarrow) [inline visual_style/views.py:21]: Yeah, seems like it should live in a utils module alongside template_finder.

**Comment by pmdarrow** [comment]
:+1:  for easy-to-read, well documented code.

File: visual_style/management/commands/update_bootstrap_examples.py (hunk 1 above)
Comment (bremac) [inline visual_style/management/commands/update_bootstrap_examples.py]: That works if we assume that we'll always serve the bootstrap docs, which doesn't seem like a given if we're not installing bootstrap through bower.

File: visual_style/views.py (hunk 6 above)
Comment (bremac) [inline visual_style/views.py:21]: Turns out that a fair amount of this method is just concerned with HTTP-specific presentation, ex. reversing URLs, generating names, etc., so I'm going to leave this in views.

**Comment by bremac** [comment]
@pmdarrow is this what you had in mind? (Aside from breaking it out into its own repo.)

**Pull request closed without merging**